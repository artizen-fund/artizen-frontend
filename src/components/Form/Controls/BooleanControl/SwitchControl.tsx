import styled from 'styled-components'
import { breakpoint, palette, Palette } from '@theme'
import { rgba } from '@lib'
import { BooleanControlProps } from './'

export const SwitchControl = ({
  data,
  label,
  uischema,
  handleChange,
  path,
  required,
  ...props
}: BooleanControlProps) => (
  <Wrapper gridArea={path} {...props}>
    <Label color={uischema?.options?.labelColor}>{typeof label === 'object' ? label[0] : label}</Label>
    <Input type="checkbox" required={!!required} onChange={_ => handleChange(path, !data)} checked={data} />
    <Switch checked={data} />
  </Wrapper>
)

// todo: investigate whether this component should use _Common/InputWrapper
//        _Common/InputLabel will not be suitable as this input is not animated.
const Wrapper = styled.label<{ gridArea?: string }>`
  position: relative;
  ${props => props.gridArea && `grid-area: ${props.gridArea};`}
  display: flex;
  justify-content: space-between;
  align-items: center;
  user-select: none;
  cursor: pointer;
  padding: 16px 0;
`

const Input = styled.input`
  position: absolute;
  opacity: 0;
  height: 0;
  width: 0;
`

const Switch = styled.span<{ checked?: boolean }>`
  appearance: none;
  position: relative;
  width: 40px;
  height: 24px;
  border-radius: 12px;

  background-color: ${props => (props.checked ? palette.white : palette.moon)};
  border: 1px solid ${props => (props.checked ? rgba(palette.night, 0.64) : rgba(palette.night, 0.32))};

  transition: background-color 0.4s ease-in-out, border-color 0.4s ease-in-out;
  will-change: background-color, border-color;

  &:after {
    content: '';
    position: absolute;
    left: 0;
    top: 3px;
    display: block;
    width: 16px;
    height: 16px;
    border-radius: 100%;

    background-color: ${props => (props.checked ? rgba(palette.uiSuccess, 0.8) : rgba(palette.night, 0.24))};
    border: 1px solid ${props => (props.checked ? rgba(palette.night, 0.32) : rgba(palette.night, 0.24))};

    transform: translateX(${props => (props.checked ? 19 : 4)}px);
    transition: background-color 0.25s ease-in-out, transform 0.4s ease-in;
    will-change: background-color, transform;
  }

  @media only screen and (min-width: ${breakpoint.laptop}px) {
    ${Wrapper}:hover & {
      border-color: ${rgba(palette.night, 0.64)};
      &:after {
        background-color: ${props => (props.checked ? rgba(palette.uiSuccess, 1) : rgba(palette.night, 0.4))};
      }
    }
  }
`

const Label = styled.span<{ color?: keyof Palette }>`
  display: block;
  max-width: 280px;

  @media only screen and (min-width: ${breakpoint.laptop}px) {
    max-width: 300px;
  }

  @media only screen and (min-width: ${breakpoint.desktop}px) {
    max-width: 320px;
  }

  color: ${props => rgba(props.color ? palette[props.color] : palette.barracuda)};
`

export default SwitchControl
