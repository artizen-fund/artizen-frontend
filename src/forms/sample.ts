export const schema = {
  type: 'object',
  properties: {
    name: {
      type: 'string',
      minLength: 3,
      description: 'Please enter your name',
    },
    vegetarian: {
      type: 'boolean',
    },
    vegetarianSure: {
      type: 'boolean',
    },
    carnivore: {
      type: 'boolean',
    },
    birthDate: {
      type: 'string',
      format: 'date',
    },
    nationality: {
      type: 'string',
      enum: ['', 'DE', 'IT', 'JP', 'US', 'RU', 'Other'],
    },
    iceCream: {
      type: 'string',
      enum: ['vanilla', 'chocolate', 'strawberry', 'neapolitan'],
    },
    personalData: {
      type: 'object',
      properties: {
        age: {
          type: 'integer',
          description: 'Please enter your age.',
        },
        height: {
          type: 'number',
        },
        drivingSkill: {
          type: 'number',
          maximum: 10,
          minimum: 1,
          default: 7,
        },
      },
      required: ['age', 'height'],
    },
    occupation: {
      type: 'string',
    },
    postalCode: {
      type: 'string',
      maxLength: 5,
    },
  },
  required: ['occupation', 'nationality'],
}

export const uischema = {
  type: 'VerticalLayout',
  elements: [
    {
      type: 'HorizontalLayout',
      elements: [
        {
          type: 'Control',
          scope: '#/properties/name',
        },
        {
          type: 'Control',
          scope: '#/properties/vegetarian',
        },
        {
          type: 'Control',
          scope: '#/properties/vegetarianSure',
          options: {
            format: 'checkbox',
            size: 'small',
          },
        },
        {
          type: 'Control',
          scope: '#/properties/carnivore',
          options: {
            format: 'switch',
          },
        },
        {
          type: 'Control',
          scope: '#/properties/iceCream',
          options: {
            format: 'segmented',
          },
        },
        {
          type: 'Control',
          scope: '#/properties/personalData/properties/age',
        },
        {
          type: 'Control',
          scope: '#/properties/birthDate',
        },
      ],
    },
    {
      type: 'Label',
      text: 'Additional Information',
    },
    {
      type: 'HorizontalLayout',
      elements: [
        {
          type: 'Control',
          scope: '#/properties/personalData/properties/height',
        },
        {
          type: 'Control',
          scope: '#/properties/nationality',
        },
        {
          type: 'Control',
          scope: '#/properties/occupation',
          suggestion: [
            'Accountant',
            'Engineer',
            'Freelancer',
            'Journalism',
            'Physician',
            'Student',
            'Teacher',
            'Other',
          ],
        },
      ],
    },
  ],
}

export const initialState = {
  name: undefined,
  vegetarian: undefined,
  vegetarianSure: undefined,
  carnivore: undefined,
  birthDate: undefined,
  iceCream: undefined,
  personalData: {
    age: undefined,
  },
  postalCode: undefined,
}
