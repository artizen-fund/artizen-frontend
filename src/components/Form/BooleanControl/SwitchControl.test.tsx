import '@testing-library/jest-dom/extend-expect'
import { render } from '@testing-library/react'
import SwitchControl from './SwitchControl'
import { schema, uischema } from '@forms/sample'

describe('SwitchControl', () => {
  const INPUT_PATH = '#/properties/switchBooleanExample'

  it('renders SelectControl unchanged', () => {
    const handleChange = () => console.log('derp')
    const { container } = render(
      <SwitchControl
        data="vanilla"
        path={uischema.elements.find(e => e.scope === INPUT_PATH)?.scope!}
        {...{ handleChange, schema, uischema }}
        label="Switch control test"
      />,
    )
    expect(container).toMatchSnapshot()
  })
})

/* 
Possible future testsâ€¦

- onchange triggers console log
*/
