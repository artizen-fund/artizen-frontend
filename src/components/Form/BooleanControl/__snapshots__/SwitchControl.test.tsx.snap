// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwitchControl renders SelectControl unchanged 1`] = `
<div>
  <label
    class="sc-6337173a-0 iDiBbZ"
  >
    <span
      class="sc-6337173a-3 efYrPI"
    >
      Switch control test
    </span>
    <input
      checked=""
      class="sc-6337173a-1 gypJCM"
      type="checkbox"
    />
    <span
      class="sc-6337173a-2 jLxQwG"
    />
  </label>
</div>
`;
